crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+            if item.1 < MAX_NUM_OF_KEYS as u8 {
+                // if an index has been set already (thus, there is a duplicate).
+                if bitvec[i] {
+                    return Err(CryptoMaterialError::BitVecError);
+                } else {
+                    sigvec[i] = item.0.clone();
+                    bitvec.set(i, true);
+                }
+            } else {
+                return Err(CryptoMaterialError::BitVecError);
+            }
+        }
+        Ok(MultiEd25519Signature(sigvec, bitvec))
+    }
+
+    /// Serialize a MultiEd25519Signature in the form of sig0||sig1||..sigN||bitvec.
+    /// Note that we omit bitvec if all bits from index = zero to index = signatures.len() - 1
+    /// are set.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes: Vec<u8> = vec![];
+        self.0.iter().for_each(|signature| {
+            bytes.extend(signature.to_bytes().to_vec());
+        });
+        // TODO compress even more, i.e., we don't always need up to 4bytes for bitvec.
+        // No need to attach the bitvector when there is no gap between the indexes.
+        if (0..self.0.len()).any(|i| !self.1[i]) {
+            bytes.extend(self.1.to_bytes());
+        }
+        bytes
+    }
+}
+
+impl TryFrom<&[u8]> for MultiEd25519Signature {
+    type Error = CryptoMaterialError;
+
+    /// Deserialize a MultiEd25519Signature. This method will also check for bitvec validity.
+    fn try_from(bytes: &[u8]) -> std::result::Result<MultiEd25519Signature, CryptoMaterialError> {
+        let length = bytes.len();
+        // Special case for single signatures to speed up deserialization.
+        if length == ED25519_SIGNATURE_LENGTH {
+            return Ed25519Signature::try_from(bytes).map(|signature| {
+                MultiEd25519Signature(vec![signature], BITVEC_FIRST_KEY_SET.clone())
+            });
+        }
+        let bitvec_num_of_bytes = length % ED25519_SIGNATURE_LENGTH;
+        let num_of_sigs = length / ED25519_SIGNATURE_LENGTH;
+
+        if num_of_sigs == 0
+            || num_of_sigs > MAX_NUM_OF_KEYS
+            || (bitvec_num_of_bytes != 0 && bitvec_num_of_bytes != BITVEC_NUM_OF_BYTES)
+        {
+            return Err(CryptoMaterialError::WrongLengthError);
+        }
+
+        let bitvec: BitVec = if bitvec_num_of_bytes == BITVEC_NUM_OF_BYTES {
+            let temp = BitVec::from_bytes(&bytes[length - 4..length]);
+            let count_set_bits = count_set_bits(&temp);
+            if count_set_bits != num_of_sigs || count_set_bits != 0 {
+                return Err(CryptoMaterialError::DeserializationError);
+            }
+            temp
+        } else {
+            let mut temp = BitVec::from_elem(MAX_NUM_OF_KEYS, false);
+            for i in 0..num_of_sigs {
+                temp.set(i, true);
+            }
+            temp
+        };
+
+        let mut signatures: Vec<Ed25519Signature> = vec![];
+        let multi_signature = bytes
```

Comment:
Same technique as suggested above: use `chunks_exact`, `map` and [collect into a `Result`](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect).